public/apps/configuration/panels/internal-user-edit/attribute-panel.tsx

```diff
@@ -0,0 +1,122 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import {
+  EuiButton,
+  EuiFieldText,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiSpacer,
+  EuiFormRow,
+} from '@elastic/eui';
+import { map } from 'lodash';
+import React, { Dispatch, Fragment, SetStateAction } from 'react';
+import { UserAttributes } from '../../types';
+import {
+  appendElementToArray,
+  removeElementFromArray,
+  updateElementInArrayHandler,
+} from '../../utils/array-state-utils';
+import { PanelWithHeader } from '../../utils/panel-with-header';
+import { UserAttributeStateClass } from './types';
+import { FormRow } from '../../utils/form-row';
+
+export function buildAttributeState(attributesDict: UserAttributes): UserAttributeStateClass[] {
+  return map(attributesDict, (v, k) => ({
+    key: k || '',
+    value: v,
+  }));
+}
+
+export function unbuildAttributeState(attributesList: UserAttributeStateClass[]): UserAttributes {
+  return attributesList.reduce(
+    (attributes: UserAttributes, { key, value }: UserAttributeStateClass) => ({
+      ...attributes,
+      [key]: value,
+    }),
+    {}
+  );
+}
+
+function getEmptyAttribute() {
+  return { key: '', value: '' };
+}
+
+function generateAttributesPanels(
+  userAttributes: UserAttributeStateClass[],
+  setAttributes: Dispatch<SetStateAction<UserAttributeStateClass[]>>
+) {
+  const panels = userAttributes.map((userAttribute, arrayIndex) => {
+    const onValueChangeHandler = (attributeToUpdate: string) =>
+      updateElementInArrayHandler(userAttributes, [arrayIndex, attributeToUpdate]);
+
+    return (
+      <Fragment key={`attributes-${arrayIndex}`}>
+        <EuiFlexGroup justifyContent="flexStart">
+          <EuiFlexItem grow={false}>
+            <FormRow headerText={arrayIndex === 0 ? 'Variable name' : ''}>
+              <EuiFieldText
+                onChange={onValueChangeHandler('key')}
+                placeholder="Type in variable name"
+              />
+            </FormRow>
+          </EuiFlexItem>
+          <EuiFlexItem grow={false}>
+            <FormRow headerText={arrayIndex === 0 ? 'Value' : ''}>
+              <EuiFieldText onChange={onValueChangeHandler('value')} placeholder="Type in value" />
+            </FormRow>
+          </EuiFlexItem>
+          <EuiFlexItem grow={false}>
+            <EuiFormRow hasEmptyLabelSpace={arrayIndex === 0 ? true : false}>
```

Comment:
Just curious why hasEmptyLabelSpace is needed. From the mockup, looks like there is header text regardless of the array index.

Reply:
It adds a spacer here for vertical alignment:
![image](https://user-images.githubusercontent.com/63078162/84716939-57378e00-af29-11ea-9185-fa6efd4f611a.png)